include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +37,92 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+using _QueueFactory = sycl::queue (*)();
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(std::pair<uintptr_t, _QueueFactory>));
+    static_assert(alignof(sycl::queue) >= alignof(std::uintptr_t));
+
+    union
+    {
+        sycl::queue __q;
+        std::pair<uintptr_t, _QueueFactory> __flag_and_factory;
+    };
+
+    bool
+    __has_queue() const
+    {
+        bool res = true;
+#if _ONEDPL_PREDEFINED_POLICIES
+        // If the first size-of-pointer bytes are zeros, we consider there is no valid queue.
+        res = (__flag_and_factory.first != 0);
+        std::atomic_signal_fence(std::memory_order_acq_rel); // to prevent possible reordering due to type punning
+#endif
+        return res;
+    }
+
+  public:
+    template <typename... _Args>
+    __queue_holder(_Args... __args) : __q(std::forward<_Args>(__args)...) {}
```

Comment:
I still do not get the second part; give an example perhaps?

Reply:
I hope this will work, and it is a better code anyway because again the purpose is clearly seen from the code itself:
```
    template<typename T, typename = std::enable_if_t<std::is_constructible_v<sycl::queue, T&&>>>
    __queue_holder(T&& __qarg) : __q(std::forward<T>(__qarg)) {}
```